File: experiment-5/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep2/src/brazilian_soccer/format.clj
;; =============================================================================
;; brazilian-soccer.format
;; -----------------------------------------------------------------------------
;; CONTEXT
;;   Part of the Brazilian Soccer MCP server (see TASK.md). Turns the structured
;;   results produced by brazilian-soccer.queries into the human-readable text
;;   blocks shown in the spec's "Example answer format" sections. The MCP layer
;;   returns both this formatted text and the raw structured data, so an LLM
;;   client can either present the prose directly or reason over the data.
;; =============================================================================
(ns brazilian-soccer.format
  (:require [clojure.string :as str]))

(defn- comp-tag [m]
  (let [parts (remove str/blank?
                      [(:competition m)
                       (when (:round m) (str "Round " (:round m)))
                       (:stage m)])]
    (when (seq parts) (str " (" (str/join ", " parts) ")"))))

(defn match-line
  "One-line description of a match, e.g.
   \"2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Série A, Round 22)\"."
  [m]
  (str (or (:date m) "????-??-??") ": "
       (:home m) " " (:home-goal m) "-" (:away-goal m) " " (:away m)
       (comp-tag m)))

(defn matches->text [ms]
  (if (empty? ms)
    "No matches found."
    (str/join "\n" (map #(str "- " (match-line %)) ms))))

(defn head-to-head->text [{:keys [team-a team-b matches record]}]
  (let [{:keys [a-wins b-wins draws total a-goals b-goals]} record]
    (if (zero? total)
      (format "No matches found between %s and %s in the dataset." team-a team-b)
      (str (format "%s vs %s — %d matches in dataset\n" team-a team-b total)
           (matches->text (take 15 matches))
           (when (> total 15) (format "\n- ... (%d more)" (- total 15)))
           (format "\n\nHead-to-head: %s %d wins, %s %d wins, %d draws (goals %d-%d)"
                   team-a a-wins team-b b-wins draws a-goals b-goals)))))

(defn team-stats->text [{:keys [team venue matches wins draws losses
                                goals-for goals-against win-rate]}]
  (if (zero? matches)
    (format "No matches found for %s with the given filters." team)
    (format "%s record%s:\n- Matches: %d\n- Wins: %d, Draws: %d, Losses: %d\n- Goals For: %d, Goals Against: %d\n- Win rate: %s%%"
            team
            (case venue :home " (home)" :away " (away)" "")
            matches wins draws losses goals-for goals-against win-rate)))

(defn standings->text [rows {:keys [competition season limit] :or {limit 20}}]
  (if (empty? rows)
    (format "No %s data found for %s." competition season)
    (str (format "%s %s standings (computed from matches):\n" season competition)
         (str/join "\n"
                   (map-indexed
                    (fn [i {:keys [team points wins draws losses goal-diff]}]
                      (format "%2d. %s - %d pts (%dW %dD %dL, GD %+d)"
                              (inc i) team points wins draws losses goal-diff))
                    (take limit rows))))))

(defn competition-stats->text [{:keys [matches total-goals avg-goals-per-match
                                       home-win-rate away-win-rate draw-rate]}]
  (format (str "Matches: %d\nTotal goals: %d\nAverage goals per match: %s\n"
               "Home win rate: %s%%  |  Away win rate: %s%%  |  Draw rate: %s%%")
          matches total-goals avg-goals-per-match
          home-win-rate away-win-rate draw-rate))

(defn biggest-wins->text [ms]
  (if (empty? ms)
    "No matches found."
    (str/join "\n"
              (map-indexed
               (fn [i m] (format "%d. %s [margin %d]" (inc i) (match-line m) (:margin m)))
               ms))))

(defn player-line [p]
  (format "%s - Overall: %s, Potential: %s, Position: %s, Club: %s, Nationality: %s, Age: %s"
          (:name p) (:overall p) (:potential p) (:position p)
          (:club p) (:nationality p) (:age p)))

(defn players->text [ps]
  (if (empty? ps)
    "No players found."
    (str/join "\n"
              (map-indexed (fn [i p] (str (inc i) ". " (player-line p))) ps))))

File: experiments/adrianco/experiment-4/runs/language=go_model=claude-opus-4-8_tooling=none/rep1/internal/soccer/format.go
// Context
// -------
// Human-readable formatters that turn query-layer structs into the text answer
// formats shown in the specification. These are used by the MCP tools/call
// handler to produce the textual content returned to the LLM client.
package soccer

import (
	"fmt"
	"strings"
)

// dateStr renders a match date in ISO form for display.
func dateStr(m Match) string { return m.Date.Format("2006-01-02") }

// FormatMatch renders one match as a single line, e.g.
// "2023-09-03: Flamengo 2-1 Fluminense (Brasileirão Série A Round 22)".
func FormatMatch(m Match) string {
	var ctx []string
	if m.Competition != "" {
		ctx = append(ctx, m.Competition)
	}
	if m.Round != "" {
		ctx = append(ctx, "Round "+m.Round)
	}
	if m.Stage != "" {
		ctx = append(ctx, m.Stage)
	}
	suffix := ""
	if len(ctx) > 0 {
		suffix = " (" + strings.Join(ctx, " ") + ")"
	}
	return fmt.Sprintf("%s: %s %d-%d %s%s", dateStr(m), m.HomeTeam, m.HomeGoals, m.AwayGoals, m.AwayTeam, suffix)
}

// FormatMatches renders a list of matches with an optional cap, appending a
// "... (N more)" line when truncated.
func FormatMatches(matches []Match, max int) string {
	if len(matches) == 0 {
		return "No matches found."
	}
	var b strings.Builder
	shown := matches
	if max > 0 && len(matches) > max {
		shown = matches[:max]
	}
	for _, m := range shown {
		fmt.Fprintf(&b, "- %s\n", FormatMatch(m))
	}
	if len(matches) > len(shown) {
		fmt.Fprintf(&b, "- ... (%d more matches in dataset)\n", len(matches)-len(shown))
	}
	return strings.TrimRight(b.String(), "\n")
}

// FormatH2H renders a head-to-head summary with a sample of recent matches.
func FormatH2H(h H2H, maxMatches int) string {
	var b strings.Builder
	fmt.Fprintf(&b, "%s vs %s:\n", h.TeamA, h.TeamB)
	if len(h.Matches) == 0 {
		b.WriteString("No matches found between these teams in the dataset.")
		return b.String()
	}
	fmt.Fprintln(&b, FormatMatches(h.Matches, maxMatches))
	fmt.Fprintf(&b, "\nHead-to-head in dataset: %s %d wins, %s %d wins, %d draws\n",
		h.TeamA, h.WinsA, h.TeamB, h.WinsB, h.Draws)
	fmt.Fprintf(&b, "Goals: %s %d, %s %d", h.TeamA, h.GoalsA, h.TeamB, h.GoalsB)
	return b.String()
}

// FormatTeamRecord renders a team's aggregate record.
func FormatTeamRecord(r TeamRecord) string {
	var title strings.Builder
	title.WriteString(r.Team)
	switch r.Filter {
	case "home":
		title.WriteString(" home record")
	case "away":
		title.WriteString(" away record")
	default:
		title.WriteString(" record")
	}
	var scope []string
	if r.Season != 0 {
		scope = append(scope, fmt.Sprintf("%d", r.Season))
	}
	if r.Competition != "" {
		scope = append(scope, r.Competition)
	}
	if len(scope) > 0 {
		title.WriteString(" (" + strings.Join(scope, " ") + ")")
	}

	var b strings.Builder
	fmt.Fprintf(&b, "%s:\n", title.String())
	fmt.Fprintf(&b, "- Matches: %d\n", r.Played)
	fmt.Fprintf(&b, "- Wins: %d, Draws: %d, Losses: %d\n", r.Wins, r.Draws, r.Losses)
	fmt.Fprintf(&b, "- Goals For: %d, Goals Against: %d\n", r.GoalsFor, r.GoalsAgainst)
	fmt.Fprintf(&b, "- Points: %d\n", r.Points())
	fmt.Fprintf(&b, "- Win rate: %.1f%%", r.WinRate())
	return b.String()
}

// FormatStandings renders a league table. limit<=0 shows all rows.
func FormatStandings(rows []TeamRecord, competition string, season int, limit int) string {
	if len(rows) == 0 {
		return "No standings available for the requested competition/season."
	}
	var b strings.Builder
	fmt.Fprintf(&b, "%d %s Final Standings (calculated from matches):\n", season, competition)
	for i, r := range rows {
		if limit > 0 && i >= limit {
			break
		}
		tag := ""
		if i == 0 {
			tag = " - Champion"
		}
		fmt.Fprintf(&b, "%d. %s - %d pts (%dW, %dD, %dL, GD %+d)%s\n",
			i+1, r.Team, r.Points(), r.Wins, r.Draws, r.Losses, r.GoalDiff(), tag)
	}
	return strings.TrimRight(b.String(), "\n")
}

// FormatPlayers renders a numbered list of players.
func FormatPlayers(players []Player, max int) string {
	if len(players) == 0 {
		return "No players found."
	}
	shown := players
	if max > 0 && len(players) > max {
		shown = players[:max]
	}
	var b strings.Builder
	for i, p := range shown {
		club := p.Club
		if club == "" {
			club = "Free agent"
		}
		fmt.Fprintf(&b, "%d. %s - Overall: %d, Position: %s, Club: %s, Nationality: %s\n",
			i+1, p.Name, p.Overall, emptyDash(p.Position), club, p.Nationality)
	}
	if len(players) > len(shown) {
		fmt.Fprintf(&b, "... (%d more players in dataset)\n", len(players)-len(shown))
	}
	return strings.TrimRight(b.String(), "\n")
}

func emptyDash(s string) string {
	if s == "" {
		return "-"
	}
	return s
}

// FormatStatistics renders competition-wide aggregate statistics.
func FormatStatistics(cs CompetitionStats) string {
	var b strings.Builder
	scope := "all competitions"
	if cs.Competition != "" {
		scope = cs.Competition
	}
	if cs.Season != 0 {
		scope = fmt.Sprintf("%s %d", scope, cs.Season)
	}
	fmt.Fprintf(&b, "Statistics for %s:\n", scope)
	fmt.Fprintf(&b, "- Matches: %d\n", cs.Matches)
	fmt.Fprintf(&b, "- Average goals per match: %.2f\n", cs.AvgGoals())
	fmt.Fprintf(&b, "- Home win rate: %.1f%% (Home %d, Away %d, Draws %d)\n",
		cs.HomeWinRate(), cs.HomeWins, cs.AwayWins, cs.Draws)
	if len(cs.BiggestWins) > 0 {
		fmt.Fprintf(&b, "\nBiggest victories:\n")
		for i, m := range cs.BiggestWins {
			fmt.Fprintf(&b, "%d. %s\n", i+1, FormatMatch(m))
		}
	}
	return strings.TrimRight(b.String(), "\n")
}
